runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -11797,6 +11797,356 @@ static TR::Register *inlineEncodeUTF16(TR::Node *node, TR::CodeGenerator *cg)
+   TR::Register *result = cg->allocateRegister();
+   TR::Register *arrAddress = cg->allocateRegister();
+   TR::Register *currentAddress = cg->allocateRegister();
+   TR::Register *endAddress = cg->allocateRegister();
+
+   TR::Register *targetVector = cg->allocateRegister(TR_VRF);
+   TR::Register *targetVectorNot = cg->allocateRegister(TR_VRF);
+   TR::Register *searchVector = cg->allocateRegister(TR_VRF);
+   TR::Register *permuteVector = cg->allocateRegister(TR_VRF);
+
+   TR_PPCScratchRegisterManager *srm = cg->generateScratchRegisterManager();
+
+   TR::LabelSymbol *startLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *notSmallLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *vectorLoopLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *residueLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *notFoundLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *foundLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *foundExactLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *endLabel = generateLabelSymbol(cg);
+
+   startLabel->setStartInternalControlFlow();
+   endLabel->setEndInternalControlFlow();
+
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, startLabel);
+
+   // Special case for empty strings, which always return -1
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::cmp4, node, cr0, startIndex, endIndex);
+   generateConditionalBranchInstruction(cg, TR::InstOpCode::beq, node, notFoundLabel, cr0);
+
+   if (!isLatin1)
+      {
+      generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, result, startIndex, startIndex);
+      generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, endAddress, endIndex, endIndex);
+      }
+   else
+      {
+      generateTrg1Src1Instruction(cg, TR::InstOpCode::mr, node, result, startIndex);
+      generateTrg1Src1Instruction(cg, TR::InstOpCode::mr, node, endAddress, endIndex);
+      }
+
+   if (node->getChild(3)->getReferenceCount() == 1)
+      srm->donateScratchRegister(startIndex);
+   if (node->getChild(4)->getReferenceCount() == 1)
+      srm->donateScratchRegister(endIndex);
+
+   generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addi, node, arrAddress, arrObjectAddress, TR::Compiler->om.contiguousArrayHeaderSizeInBytes());
+   if (node->getChild(1)->getReferenceCount() == 1)
+      srm->donateScratchRegister(arrObjectAddress);
+
+   // Handle the first character using a simple scalar compare. Otherwise, first character matches
+   // would be slower than the old scalar comparison loop. This is a problem since first character
+   // matches are common in certain contexts, e.g. StringTokenizer where the default first character
+   // to each String.indexOf call is ' ', which is the most common whitespace character.
+      {
+      TR::Register *value = srm->findOrCreateScratchRegister();
+
+      generateTrg1MemInstruction(cg, scalarLoadOp, node, value, new (cg->trHeapMemory()) TR::MemoryReference(arrAddress, result, isLatin1 ? 1 : 2, cg));
```

Comment:
Please switch the base/index positions of arrAddress and result registers in this MemoryReference constructor, since arrAddress register is not set excludingGPR0 but result register is.

Reply:
Fixed